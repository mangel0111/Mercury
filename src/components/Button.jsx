import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const ButtonBase = styled.button`
    display: inline-block;
    font-weight: 400;
    text-align: center;
    white-space: nowrap;
    vertical-align: middle;
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
    border: 1px solid transparent;
    padding: .375rem .75rem;
    font-size: 1rem;
    line-height: 1.5;
    border-radius: .25rem;
    transition: color .15s ease-in-out,background-color .15s ease-in-out,border-color .15s ease-in-out,box-shadow .15s ease-in-out;    
    color: #fff;
    background-color: ${props => bgColorChooser(props.type)};
    border-color: ${props => bgColorChooser(props.type)};
`;

const bgColorChooser = (type) => {
	switch(type){
	case 'PRIMARY': 
		return  '#28a745';
	case 'SECONDARY':
		return '#6c757d';
	case 'DANGER': 
		return '#dc3545';
	case 'NEUTRAL':
	default: 
		return '#f8f9fa';
	}
};

class Button extends Component {
	render(){
		const { label, onClick, type = 'NEUTRAL'} = this.props;
		return (
			<ButtonBase
				type={type}
				onClick={onClick}
			>
				{label}
			</ButtonBase>
		);
	}
}

Button.propTypes = {
	label: PropTypes.string,
	onClick: PropTypes.func,
	type: PropTypes.string
};

export default Button;
