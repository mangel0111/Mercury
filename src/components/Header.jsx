import React, { Component } from 'react';
import styled from 'styled-components';

const HeaderNav = styled.header`
    padding: 0;
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 55px;
    z-index: 2;
    background-color: #000;
    color: #fff;
`;

const Container = styled.div`
    margin-right: 10%;
    margin-left: 10%;
    padding-left: 15px;
    padding-right: 15px;
`;

const Nav = styled.div`
    display: flex;
    margin: 10px;
    font-size: 20px;
    font-weight: lighter;
`;

const Menu = styled.div`
    margin-left: auto;
`;

export class Header extends Component{
	render (){
		return (
			<HeaderNav>
				<Container>
					<Nav>
						<div>
                            Mercury
						</div>
						<Menu>
                            Home
						</Menu>
					</Nav>
				</Container>
			</HeaderNav>
		);
	}
}

export default Header;
