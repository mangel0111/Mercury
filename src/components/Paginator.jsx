import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Navigator = styled.nav`
    display: flex;
`;

const PageNavigator =  styled.ul`
    margin-left: auto;
    display: flex;
    padding-left: 0;
    list-style: none;
    border-radius: .25rem;
    margin-top: .5rem;
    margin-bottom: .5rem;

    a {
        cursor: pointer;
    }
`;

const Link = styled.a`
    position: relative;
    display: block;
    padding: .5rem .75rem;
    margin-left: -1px;
    line-height: 1.25;
    color: ${props => props.selected ? '#fff': '#007bff'};
    background-color: ${props => props.selected ? '#007bff': '#fff'};;
    border: 1px solid #dee2e6;
    margin-left: 0;
`;

const FirtsLink = styled(Link)`
    border-top-left-radius: .25rem;
    border-bottom-left-radius: .25rem;
`;

const LastLink = styled(Link)`
    border-top-right-radius: .25rem;
    border-bottom-right-radius: .25rem;

`;

export class Paginator extends Component {
    
	getMidlePages() {
		const { pageSize, size } = this.props;
		let maxPages = Math.ceil(size / pageSize);
		if(maxPages <= 2){
			return [];
		}
		const pages = Array.from(new Array(--maxPages),(val,index)=>index);
		return pages.slice(1);
	}

	render(){
		const { pageSize, size, onChange, page } = this.props;
    
		if(!size || size <= pageSize) {
			return null;
		}
		const currentPage = Math.ceil(size / pageSize);
		return (
			<Navigator>
				<PageNavigator>
					<li>
						<FirtsLink
							selected={page===0}
							onClick={()=> onChange({page: 0})}
						>
                            Previous
						</FirtsLink>
					</li>
					{this.getMidlePages().map(pageItem => 
						(
							<li 
								key={pageItem}
							>
								<Link
									selected={pageItem === page}
									onClick={()=> onChange({page: --pageItem})}s
								>
									{++pageItem}
								</Link>
							</li>
						)
					)}
					<li>
						<LastLink
							selected={currentPage - 1 === page}
							onClick={()=> onChange({page: currentPage - 1})}
						>
                            Last
						</LastLink>
					</li>
				</PageNavigator>
			</Navigator>
		);
	}
}

Paginator.propTypes = {
	onChange: PropTypes.func,
	size: PropTypes.number,
	pageSize: PropTypes.number,
	page: PropTypes.number
};

export default Paginator;
