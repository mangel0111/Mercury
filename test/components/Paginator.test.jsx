import 'jsdom-global/register'; 
import React from 'react';
import Adapter from 'enzyme-adapter-react-15';
import renderer from 'react-test-renderer';
import  Enzyme, { mount } from 'enzyme';
import { Paginator } from '../../src/components/Paginator';

let props = {
	page: 0,
	size: 10,
	pageSize: 5,
	onChange: jest.fn()
};

describe('Renders Paginator', () => {
	beforeEach(()=> {
		Enzyme.configure({ adapter: new Adapter() });
		props = {
			page: 0,
			size: 10,
			pageSize: 5,
			onChange: jest.fn()
		};
	});
	it('render without crash', ()=> {
		const rendered = renderer.create(<Paginator {...props} />);
		const tree = rendered.toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('render without crash', ()=> {
		props = {
			page: 0,
			size: 50,
			pageSize: 5,
			onChange: jest.fn()
		};
		const rendered = renderer.create(<Paginator {...props} />);
		const tree = rendered.toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('should render the correct number of links for 2 pages', ()=> {
		let paginator = mount(<Paginator {...props} />);
		let links = paginator.find('li');
		expect(links.length).toEqual(2);
	});
	it('should render the correct number of links for 5 pages',()=> {
		props = {
			page: 0,
			size: 50,
			pageSize: 5,
			onChange: jest.fn()
		};
        
		const paginator = mount(<Paginator {...props} />);
		const links = paginator.find('li');
		expect(links.length).toEqual(10);
	});
    
	it('should paginate correclty and go to the first link', ()=> {
		props = {
			page: 0,
			size: 50,
			pageSize: 5,
			onChange: jest.fn()
		};
		const paginator = mount(<Paginator {...props} />);

		const firstLink = paginator.find('[data-action="paginator:first:link"]');
		firstLink.at(0).props().onClick();
		expect(props.onChange.mock.calls.length).toEqual(1);
		expect(props.onChange.mock.calls[0][0]).toEqual(
			{'page': 0}
		);
	});
	it('should paginate correclty and go to the last link', ()=> {
		props = {
			page: 0,
			size: 50,
			pageSize: 5,
			onChange: jest.fn()
		};
		const paginator = mount(<Paginator {...props} />);
        
		const lastLink = paginator.find('[data-action="paginator:last:link"]');
		lastLink.at(0).props().onClick();
		expect(props.onChange.mock.calls.length).toEqual(1);
		expect(props.onChange.mock.calls[0][0]).toEqual(
			{'page': 9}
		);
	});
	it('should paginate correclty and go to the others links (not first netiher last)', ()=> {
		props = {
			page: 0,
			size: 50,
			pageSize: 5,
			onChange: jest.fn()
		};
		const paginator = mount(<Paginator {...props} />);
        
		const pages = Array.from(new Array(8),(val,index)=>index).slice(1);
		pages.forEach((page, index) => {
            
			const currentPage = index + 2;
			const pageLink = paginator.find(`[data-action="paginator:${currentPage}:link"]`);
			pageLink.at(0).props().onClick();
			expect(props.onChange.mock.calls.length).toEqual(index + 1);
			expect(props.onChange.mock.calls[index][0]).toEqual(
				{'page': currentPage}
			);
		});
	});
});
