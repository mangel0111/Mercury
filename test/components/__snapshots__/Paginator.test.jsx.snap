// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Paginator render without crash 1`] = `
<nav
  className="sc-bdVaJa iSsrAY"
  data-container="paginator"
>
  <ul
    className="sc-bwzfXH jZnbsg"
  >
    <li>
      <a
        className="sc-bxivhb hJoAke sc-htpNat hJqWET"
        data-action="paginator:first:link"
        onClick={[Function]}
        selected={true}
      >
        Previous
      </a>
    </li>
    <li>
      <a
        className="sc-ifAKCX hJgFXe sc-htpNat ioIdIC"
        data-action="paginator:last:link"
        onClick={[Function]}
        selected={false}
      >
        Last
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Renders Paginator render without crash 2`] = `
<nav
  className="sc-bdVaJa iSsrAY"
  data-container="paginator"
>
  <ul
    className="sc-bwzfXH jZnbsg"
  >
    <li>
      <a
        className="sc-bxivhb hJoAke sc-htpNat hJqWET"
        data-action="paginator:first:link"
        onClick={[Function]}
        selected={true}
      >
        Previous
      </a>
    </li>
    <li>
      <a
        className="sc-htpNat ioIdIC"
        data-action="paginator:1:link"
        onClick={[Function]}
        selected={false}
      >
        2
      </a>
    </li>
    <li>
      <a
        className="sc-htpNat ioIdIC"
        data-action="paginator:2:link"
        onClick={[Function]}
        selected={false}
      >
        3
      </a>
    </li>
    <li>
      <a
        className="sc-htpNat ioIdIC"
        data-action="paginator:3:link"
        onClick={[Function]}
        selected={false}
      >
        4
      </a>
    </li>
    <li>
      <a
        className="sc-htpNat ioIdIC"
        data-action="paginator:4:link"
        onClick={[Function]}
        selected={false}
      >
        5
      </a>
    </li>
    <li>
      <a
        className="sc-htpNat ioIdIC"
        data-action="paginator:5:link"
        onClick={[Function]}
        selected={false}
      >
        6
      </a>
    </li>
    <li>
      <a
        className="sc-htpNat ioIdIC"
        data-action="paginator:6:link"
        onClick={[Function]}
        selected={false}
      >
        7
      </a>
    </li>
    <li>
      <a
        className="sc-htpNat ioIdIC"
        data-action="paginator:7:link"
        onClick={[Function]}
        selected={false}
      >
        8
      </a>
    </li>
    <li>
      <a
        className="sc-htpNat ioIdIC"
        data-action="paginator:8:link"
        onClick={[Function]}
        selected={false}
      >
        9
      </a>
    </li>
    <li>
      <a
        className="sc-ifAKCX hJgFXe sc-htpNat ioIdIC"
        data-action="paginator:last:link"
        onClick={[Function]}
        selected={false}
      >
        Last
      </a>
    </li>
  </ul>
</nav>
`;
